File: src/wrappers/themis/jsthemis/secure_session.cpp
/*
 * Copyright (c) 2015 Cossack Labs Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "secure_session.hpp"

#include <cstring>

#include <node_buffer.h>

#include <themis/themis.h>

#include "errors.hpp"
#include "secure_keygen.hpp"

namespace jsthemis
{

Nan::Persistent<v8::Function> SecureSession::constructor;

int SecureSession::get_public_key_for_id_callback(
    const void* id, size_t id_length, void* key_buffer, size_t key_buffer_length, void* user_data)
{
    if (!key_buffer || !key_buffer_length) {
        return THEMIS_BUFFER_TOO_SMALL;
    }
    v8::Local<v8::Value> argv[1] = {Nan::CopyBuffer((char*)id, id_length).ToLocalChecked()};
    Nan::MaybeLocal<v8::Value> result =
        Nan::Call(reinterpret_cast<SecureSession*>(user_data)->id_to_pub_key_callback_, 1, argv);
    if (result.IsEmpty()) {
        return THEMIS_FAIL;
    }
    v8::Local<v8::Value> a = result.ToLocalChecked();
    if (a->IsUint8Array()) {
        v8::Local<v8::Object> buffer = a.As<v8::Object>();
        if (key_buffer_length < node::Buffer::Length(buffer)) {
            return THEMIS_BUFFER_TOO_SMALL;
        }
        std::memcpy(key_buffer,
                    (const uint8_t*)(node::Buffer::Data(buffer)),
                    node::Buffer::Length(buffer));
        return THEMIS_SUCCESS;
    }
    if (a->IsNull() || a->IsUndefined()) {
        return THEMIS_FAIL;
    }
    return THEMIS_INVALID_PARAMETER;
}

SecureSession::SecureSession(const std::vector<uint8_t>& id,
                             const std::vector<uint8_t>& private_key,
                             v8::Local<v8::Function> get_pub_by_id_callback)
    : session_(NULL)
    , id_to_pub_key_callback_(get_pub_by_id_callback)
{
    callback_.get_public_key_for_id = SecureSession::get_public_key_for_id_callback;
    callback_.send_data = NULL;
    callback_.receive_data = NULL;
    callback_.state_changed = NULL;
    callback_.user_data = this;
    session_ = secure_session_create(&id[0], id.size(), &private_key[0], private_key.size(), &callback_);
}

SecureSession::~SecureSession()
{
    if (!session_) {
        secure_session_destroy(session_);
        session_ = NULL;
    }
}

void SecureSession::Init(v8::Local<v8::Object> exports)
{
    v8::Local<v8::String> className = Nan::New("SecureSession").ToLocalChecked();
    // Prepare constructor template
    v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(New);
    tpl->SetClassName(className);
    tpl->InstanceTemplate()->SetInternalFieldCount(1);
    // Prototype
    Nan::SetPrototypeMethod(tpl, "connectRequest", connectRequest);
    Nan::SetPrototypeMethod(tpl, "wrap", wrap);
    Nan::SetPrototypeMethod(tpl, "unwrap", unwrap);
    Nan::SetPrototypeMethod(tpl, "isEstablished", isEstablished);
    // Export constructor
    v8::Local<v8::Function> function = Nan::GetFunction(tpl).ToLocalChecked();
    constructor.Reset(function);
    Nan::Set(exports, className, function);
}

void SecureSession::New(const Nan::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.IsConstructCall()) {
        if (args.Length() < 3) {
            ThrowParameterError("Secure Session constructor",
                                "not enough arguments, expected client ID, private key, public key callback");
            args.GetReturnValue().SetUndefined();
            return;
        }
        if (!args[0]->IsUint8Array()) {
            ThrowTypeError("SecureSession",
                           "client ID is not a byte buffer, use ByteBuffer or Uint8Array");
            args.GetReturnValue().SetUndefined();
            return;
        }
        if (node::Buffer::Length(args[0]) == 0) {
            ThrowParameterError("Secure Session constructor", "client ID is empty");
            args.GetReturnValue().SetUndefined();
            return;
        }
        if (!args[1]->IsUint8Array()) {
            ThrowTypeError("SecureSession",
                           "private key is not a byte buffer, use ByteBuffer or Uint8Array");
            args.GetReturnValue().SetUndefined();
            return;
        }
        if (node::Buffer::Length(args[1]) == 0) {
            ThrowParameterError("Secure Session constructor", "private key is empty");
            args.GetReturnValue().SetUndefined();
            return;
        }
        if (!args[2]->IsFunction()) {
            ThrowParameterError("Secure Session constructor", "public key callback is not a function");
            args.GetReturnValue().SetUndefined();
            return;
        }
        std::vector<uint8_t> id((uint8_t*)(node::Buffer::Data(args[0])),
                                (uint8_t*)(node::Buffer::Data(args[0]) + node::Buffer::Length(args[0])));
        std::vector<uint8_t> private_key((uint8_t*)(node::Buffer::Data(args[1])),
                                         (uint8_t*)(node::Buffer::Data(args[1])
                                                    + node::Buffer::Length(args[1])));
        if (!IsPrivateKey(private_key)) {
            ThrowParameterError("SecureSession", "invalid private key");
            args.GetReturnValue().SetUndefined();
            return;
        }
        SecureSession* obj = new SecureSession(id, private_key, v8::Local<v8::Function>::Cast(args[2]));
        if (!obj->isCreated()) {
            // secure_session_create() inside SecureSession::SecureSession()
            // may fail, we need to catch this and throw exception to JS
            ThrowParameterError("SecureSession", "unsupported private key");
            args.GetReturnValue().SetUndefined();
            return;
        }
        obj->Wrap(args.This());
        args.GetReturnValue().Set(args.This());
    } else {
        const int argc = 3;
        v8::Local<v8::Value> argv[argc] = {args[0], args[1], args[2]};
        v8::Local<v8::Function> cons = Nan::New<v8::Function>(constructor);
        args.GetReturnValue().Set(Nan::NewInstance(cons, argc, argv).ToLocalChecked());
    }
}

void SecureSession::connectRequest(const Nan::FunctionCallbackInfo<v8::Value>& args)
{
    themis_status_t status = THEMIS_FAIL;
    SecureSession* obj = Nan::ObjectWrap::Unwrap<SecureSession>(args.This());
    size_t length = 0;
    status = secure_session_generate_connect_request(obj->session_, NULL, &length);
    if (status != THEMIS_BUFFER_TOO_SMALL) {
        ThrowSecureSessionError("Secure Session failed to generate connect request", status);
        args.GetReturnValue().SetUndefined();
        return;
    }
    uint8_t* data = (uint8_t*)(malloc(length));
    status = secure_session_generate_connect_request(obj->session_, data, &length);
    if (status != THEMIS_SUCCESS) {
        ThrowSecureSessionError("Secure Session failed to generate connect request", status);
        free(data);
        args.GetReturnValue().SetUndefined();
        return;
    }
    args.GetReturnValue().Set(Nan::NewBuffer((char*)(data), length).ToLocalChecked());
}

void SecureSession::wrap(const Nan::FunctionCallbackInfo<v8::Value>& args)
{
    themis_status_t status = THEMIS_FAIL;
    SecureSession* obj = Nan::ObjectWrap::Unwrap<SecureSession>(args.This());
    if (args.Length() < 1) {
        ThrowParameterError("Secure Session failed to encrypt",
                            "not enough arguments, expected message");
        args.GetReturnValue().SetUndefined();
        return;
    }
    if (!args[0]->IsUint8Array()) {
        ThrowTypeError("SecureSession", "message is not a byte buffer, use ByteBuffer or Uint8Array");
        args.GetReturnValue().SetUndefined();
        return;
    }
    if (node::Buffer::Length(args[0]) == 0) {
        ThrowParameterError("Secure Session failed to encrypt", "message is empty");
        args.GetReturnValue().SetUndefined();
        return;
    }
    size_t length = 0;
    status = secure_session_wrap(obj->session_,
                                 (const uint8_t*)(node::Buffer::Data(args[0])),
                                 node::Buffer::Length(args[0]),
                                 NULL,
                                 &length);
    if (status != THEMIS_BUFFER_TOO_SMALL) {
        ThrowSecureSessionError("Secure Session failed to encrypt", status);
        args.GetReturnValue().SetUndefined();
        return;
    }
    uint8_t* data = (uint8_t*)(malloc(length));
    status = secure_session_wrap(obj->session_,
                                 (const uint8_t*)(node::Buffer::Data(args[0])),
                                 node::Buffer::Length(args[0]),
                                 data,
                                 &length);
    if (status != THEMIS_SUCCESS) {
        ThrowSecureSessionError("Secure Session failed to encrypt", status);
        free(data);
        args.GetReturnValue().SetUndefined();
        return;
    }
    args.GetReturnValue().Set(Nan::NewBuffer((char*)(data), length).ToLocalChecked());
}

void SecureSession::unwrap(const Nan::FunctionCallbackInfo<v8::Value>& args)
{
    themis_status_t status = THEMIS_FAIL;
    SecureSession* obj = Nan::ObjectWrap::Unwrap<SecureSession>(args.This());
    if (args.Length() < 1) {
        ThrowParameterError("Secure Session failed to decrypt",
                            "not enough arguments, expected message");
        args.GetReturnValue().SetUndefined();
        return;
    }
    if (!args[0]->IsUint8Array()) {
        ThrowTypeError("SecureSession", "message is not a byte buffer, use ByteBuffer or Uint8Array");
        args.GetReturnValue().SetUndefined();
        return;
    }
    if (node::Buffer::Length(args[0]) == 0) {
        ThrowParameterError("Secure Session failed to decrypt", "message is empty");
        args.GetReturnValue().SetUndefined();
        return;
    }
    size_t length = 0;
    status = secure_session_unwrap(obj->session_,
                                   (const uint8_t*)(node::Buffer::Data(args[0])),
                                   node::Buffer::Length(args[0]),
                                   NULL,
                                   &length);
    if (status != THEMIS_BUFFER_TOO_SMALL) {
        if (status != THEMIS_SUCCESS) {
            ThrowSecureSessionError("Secure Session failed to decrypt", status);
        }
        args.GetReturnValue().SetUndefined();
        return;
    }
    uint8_t* data = (uint8_t*)(malloc(length));
    status = secure_session_unwrap(obj->session_,
                                   (const uint8_t*)(node::Buffer::Data(args[0])),
                                   node::Buffer::Length(args[0]),
                                   data,
                                   &length);
    if (status != THEMIS_SUCCESS && status != THEMIS_SSESSION_SEND_OUTPUT_TO_PEER) {
        ThrowSecureSessionError("Secure Session failed to decrypt", status);
        free(data);
        args.GetReturnValue().SetUndefined();
        return;
    }
    args.GetReturnValue().Set(Nan::NewBuffer((char*)(data), length).ToLocalChecked());
}

void SecureSession::isEstablished(const Nan::FunctionCallbackInfo<v8::Value>& args)
{
    SecureSession* obj = Nan::ObjectWrap::Unwrap<SecureSession>(args.This());
    args.GetReturnValue().Set(Nan::New<v8::Boolean>(secure_session_is_established(obj->session_)));
}

} // namespace jsthemis

File: src/wrappers/themis/wasm/src/secure_session.js
// Copyright (c) 2019 Cossack Labs Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @file
 * Themis Secure Session.
 */

const libthemis = require('./libthemis.js')
const keygen = require('./secure_keygen.js')
const errors = require('./themis_error.js')
const utils = require('./utils.js')

const cryptosystem_name = 'SecureSession'

const ThemisError = errors.ThemisError
const ThemisErrorCode = errors.ThemisErrorCode

function isFunction(obj) {
    return !!(obj && obj.constructor && obj.call && obj.apply)
}

// Secure Session C API operatates with "secure_session_user_callbacks_t" context structure
// defined in <themis/secure_session.h>. It must be filled in with C function pointers and
// an arbitrary C context pointer. Associated C functions are called by Secure Session
// at appropriate moments. They receive the C context pointer as their last argument which
// allows to restore context.
//
// First, we need to be able to allocate and free "secure_session_user_callbacks_t", and to
// initialize it before passing them to Secure Session constructor.
//
// Here's how it looks in C:
//
//     struct secure_session_user_callbacks_type {
//          send_protocol_data_callback     send_data;
//          receive_protocol_data_callback  receive_data;
//          protocol_state_changed_callback state_changed;
//          get_public_key_for_id_callback  get_public_key_for_id;
//
//          void *user_data;
//     };
//
// On Emscripten target all pointers take up 4 bytes and are aligned at 4-byte boundary.
// Hence the following sizes and offsets:

const sizeof_secure_session_user_callbacks_t = 20
const offsetof_user_data                     = 16
const offsetof_get_public_key_for_id         = 12

// One does not simply pass JavaScript function into Emscripten memory. We need to use
// the "addFunction" API which registers a limited number of JavaScript functions and
// provides corresponding C function pointers to thunks. We need only one callback so
// it's relatively easy for us. We just need to ensure that the function is registered
// only once.
//
// As for the context, we'll save a pointer to "secure_session_user_callbacks_t" itself
// in its "user_data" field.

var getPublicKeyForIdThunkPtr = 0

function initUserCallbacks(callbacksPtr) {
    if (getPublicKeyForIdThunkPtr == 0) {
        // LLVM needs to know the prototype of the corresponding C function.
        // See comment for "getPublicKeyForIdThunk" below.
        getPublicKeyForIdThunkPtr = libthemis.addFunction(getPublicKeyForIdThunk, 'iiiiii')
    }
    libthemis._memset(callbacksPtr, 0, sizeof_secure_session_user_callbacks_t)
    // We're writing LLVM pointers into memory, hence the '*' type specifier
    libthemis.setValue(callbacksPtr + offsetof_get_public_key_for_id, getPublicKeyForIdThunkPtr, '*')
    libthemis.setValue(callbacksPtr + offsetof_user_data, callbacksPtr, '*')
}

// Now we need to be able to identify JavaScript Secure Session object associated with
// a particular "secure_session_user_callbacks_t" instance. Unfortunately, keeping another
// global registry in JavaScript is the only feasible choice because JavaScript object
// memory is managed completely opaquely.

var SecureSessionCallbackRegistry = {}

function registerSecureSession(callbacksPtr, session) {
    SecureSessionCallbackRegistry[callbacksPtr] = session
}

function unregisterSecureSession(callbacksPtr) {
    delete SecureSessionCallbackRegistry[callbacksPtr]
}

function getSecureSession(callbacksPtr) {
    return SecureSessionCallbackRegistry[callbacksPtr]
}

// Finally, here's our JavaScript implementation of the C function callback.
// It has the following prototype:
//
//     int get_public_key_for_id_callback(
//         const void* id,   size_t id_length,
//         void* key_buffer, size_t key_buffer_length,
//         void* user_data
//     );
//
// It is expected to look up a public key corresponding to the provided session ID,
// write the key into provided buffer, and return zero on success. Non-zero return
// values are considered errors.
//
// The function accepts five integer argumentes and returns an integer. Hence its
// LLVM signature is "iiiiii".
//
// We can throw JavaScript exceptions from inside the function, thanks to Emscripten.

const GetPublicKeySuccess = 0
const GetPublicKeyFailure = -1

function getPublicKeyForIdThunk(idPtr, idLen, keyPtr, keyLen, userData) {
    let id = libthemis.HEAPU8.slice(idPtr, idPtr + idLen)
    let session = getSecureSession(userData)

    let publicKey = session.keyCallback(id)
    if (publicKey == null) {
        return GetPublicKeyFailure
    }

    if (!(publicKey instanceof keygen.PublicKey)) {
        throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
            'Secure Session callback must return PublicKey or null')
    }
    if (publicKey.length > keyLen) {
        throw new ThemisError(cryptosystem_name, ThemisErrorCode.BUFFER_TOO_SMALL,
            'public key cannot fit into provided buffer')
    }
    utils.heapPutArray(publicKey, keyPtr)

    return GetPublicKeySuccess
}

class SecureSession {
    constructor(sessionID, privateKey, keyCallback) {
        sessionID = utils.coerceToBytes(sessionID)
        if (sessionID.length == 0) {
            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
                'session ID must be not empty')
        }
        if (!(privateKey instanceof keygen.PrivateKey)) {
            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
                'expected PrivateKey as second argument')
        }
        if (!isFunction(keyCallback)) {
            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
                'expected callback as third argument')
        }

        let session_id_ptr, private_key_ptr, callbacks_ptr
        try {
            session_id_ptr = utils.heapAlloc(sessionID.length)
            private_key_ptr = utils.heapAlloc(privateKey.length)
            callbacks_ptr = utils.heapAlloc(sizeof_secure_session_user_callbacks_t)
            if (!session_id_ptr || !private_key_ptr || !callbacks_ptr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
            }

            utils.heapPutArray(sessionID, session_id_ptr)
            utils.heapPutArray(privateKey, private_key_ptr)

            this.keyCallback = keyCallback
            initUserCallbacks(callbacks_ptr)

            this.sessionPtr = libthemis._secure_session_create(
                session_id_ptr, sessionID.length,
                private_key_ptr, privateKey.length,
                callbacks_ptr
            )
            if (!this.sessionPtr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY,
                    'failed to allocate Secure Session')
            }

            registerSecureSession(callbacks_ptr, this)
            this.callbacksPtr = callbacks_ptr
            callbacks_ptr = null
        }
        finally {
            utils.heapFree(private_key_ptr, privateKey.length)
            utils.heapFree(session_id_ptr, sessionID.length)
            utils.heapFree(callbacks_ptr, sizeof_secure_session_user_callbacks_t)
        }
    }

    destroy() {
        let status = libthemis._secure_session_destroy(this.sessionPtr)
        if (status != ThemisErrorCode.SUCCESS) {
            throw new ThemisError(cryptosystem_name, status,
                'failed to destroy Secure Session')
        }
        this.sessionPtr = null

        unregisterSecureSession(this.callbacksPtr)
        utils.heapFree(this.callbacksPtr, sizeof_secure_session_user_callbacks_t)
        this.callbacksPtr = null
    }

    established() {
        let value = libthemis._secure_session_is_established(this.sessionPtr)
        return !(value == 0)
    }

    connectionRequest() {
        let status
        /// C API uses "size_t" for lengths, it's defined as "i32" in Emscripten
        let request_length_ptr = libthemis.allocate(4, 'i32', libthemis.ALLOC_STACK)
        let request_ptr, request_length
        try {
            status = libthemis._secure_session_generate_connect_request(
                this.sessionPtr,
                null, request_length_ptr
            )
            if (status != ThemisErrorCode.BUFFER_TOO_SMALL) {
                throw new ThemisError(cryptosystem_name, status)
            }

            request_length = libthemis.getValue(request_length_ptr, 'i32')
            request_ptr = utils.heapAlloc(request_length)
            if (!request_ptr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
            }

            status = libthemis._secure_session_generate_connect_request(
                this.sessionPtr,
                request_ptr, request_length_ptr
            )
            if (status != ThemisErrorCode.SUCCESS) {
                throw new ThemisError(cryptosystem_name, status)
            }

            request_length = libthemis.getValue(request_length_ptr, 'i32')

            return utils.heapGetArray(request_ptr, request_length)
        }
        finally {
            utils.heapFree(request_ptr, request_length)
        }
    }

    negotiateReply(message) {
        message = utils.coerceToBytes(message)
        if (message.length == 0) {
            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
                'message must be not empty')
        }

        let status
        /// C API uses "size_t" for lengths, it's defined as "i32" in Emscripten
        let reply_length_ptr = libthemis.allocate(4, 'i32', libthemis.ALLOC_STACK)
        let message_ptr, reply_ptr, reply_length
        try {
            message_ptr = utils.heapAlloc(message.length)
            if (!message_ptr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
            }

            utils.heapPutArray(message, message_ptr)

            status = libthemis._secure_session_unwrap(
                this.sessionPtr,
                message_ptr, message.length,
                null, reply_length_ptr
            )
            if (status == ThemisErrorCode.SUCCESS) {
                return new Uint8Array()
            }
            if (status != ThemisErrorCode.BUFFER_TOO_SMALL) {
                throw new ThemisError(cryptosystem_name, status)
            }

            reply_length = libthemis.getValue(reply_length_ptr, 'i32')
            reply_ptr = utils.heapAlloc(reply_length)
            if (!reply_ptr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
            }

            status = libthemis._secure_session_unwrap(
                this.sessionPtr,
                message_ptr, message.length,
                reply_ptr, reply_length_ptr
            )
            if (status != ThemisErrorCode.SSESSION_SEND_OUTPUT_TO_PEER) {
                throw new ThemisError(cryptosystem_name, status)
            }

            reply_length = libthemis.getValue(reply_length_ptr, 'i32')

            return utils.heapGetArray(reply_ptr, reply_length)
        }
        finally {
            utils.heapFree(message_ptr, message.length)
            utils.heapFree(reply_ptr, reply_length)
        }
    }

    wrap(message) {
        message = utils.coerceToBytes(message)
        if (message.length == 0) {
            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
                'message must be not empty')
        }

        let status
        /// C API uses "size_t" for lengths, it's defined as "i32" in Emscripten
        let wrapped_length_ptr = libthemis.allocate(4, 'i32', libthemis.ALLOC_STACK)
        let message_ptr, wrapped_ptr, wrapped_length
        try {
            message_ptr = utils.heapAlloc(message.length)
            if (!message_ptr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
            }

            utils.heapPutArray(message, message_ptr)

            status = libthemis._secure_session_wrap(
                this.sessionPtr,
                message_ptr, message.length,
                null, wrapped_length_ptr
            )
            if (status != ThemisErrorCode.BUFFER_TOO_SMALL) {
                throw new ThemisError(cryptosystem_name, status)
            }

            wrapped_length = libthemis.getValue(wrapped_length_ptr, 'i32')
            wrapped_ptr = utils.heapAlloc(wrapped_length)
            if (!wrapped_ptr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
            }

            status = libthemis._secure_session_wrap(
                this.sessionPtr,
                message_ptr, message.length,
                wrapped_ptr, wrapped_length_ptr
            )
            if (status != ThemisErrorCode.SUCCESS) {
                throw new ThemisError(cryptosystem_name, status)
            }

            wrapped_length = libthemis.getValue(wrapped_length_ptr, 'i32')

            return utils.heapGetArray(wrapped_ptr, wrapped_length)
        }
        finally {
            utils.heapFree(message_ptr, message.length)
            utils.heapFree(wrapped_ptr, wrapped_length)
        }
    }

    unwrap(message) {
        message = utils.coerceToBytes(message)
        if (message.length == 0) {
            throw new ThemisError(cryptosystem_name, ThemisErrorCode.INVALID_PARAMETER,
                'message must be not empty')
        }

        let status
        /// C API uses "size_t" for lengths, it's defined as "i32" in Emscripten
        let unwrapped_length_ptr = libthemis.allocate(4, 'i32', libthemis.ALLOC_STACK)
        let message_ptr, unwrapped_ptr, unwrapped_length
        try {
            message_ptr = utils.heapAlloc(message.length)
            if (!message_ptr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
            }

            utils.heapPutArray(message, message_ptr)

            status = libthemis._secure_session_unwrap(
                this.sessionPtr,
                message_ptr, message.length,
                null, unwrapped_length_ptr
            )
            if (status != ThemisErrorCode.BUFFER_TOO_SMALL) {
                throw new ThemisError(cryptosystem_name, status)
            }

            unwrapped_length = libthemis.getValue(unwrapped_length_ptr, 'i32')
            unwrapped_ptr = utils.heapAlloc(unwrapped_length)
            if (!unwrapped_ptr) {
                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
            }

            status = libthemis._secure_session_unwrap(
                this.sessionPtr,
                message_ptr, message.length,
                unwrapped_ptr, unwrapped_length_ptr
            )
            if (status != ThemisErrorCode.SUCCESS) {
                throw new ThemisError(cryptosystem_name, status)
            }

            unwrapped_length = libthemis.getValue(unwrapped_length_ptr, 'i32')

            return utils.heapGetArray(unwrapped_ptr, unwrapped_length)
        }
        finally {
            utils.heapFree(message_ptr, message.length)
            utils.heapFree(unwrapped_ptr, unwrapped_length)
        }
    }
}

module.exports.SecureSession = SecureSession
